actionpack/lib/action_dispatch/testing/request_encoder.rb

```diff
@@ -53,6 +53,6 @@ def self.register_encoder(mime_name, param_encoder: nil, response_parser: nil)
     end
 
     register_encoder :html, response_parser: -> body { Rails::Dom::Testing.html_document.parse(body) }
-    register_encoder :json, response_parser: -> body { JSON.parse(body) }
+    register_encoder :json, response_parser: -> body { JSON.parse(body, object_class: ActiveSupport::HashWithIndifferentAccess) }
```

Comment:
This change is potentially breaking and not backwards compatible.

Since it's a test harness change, and doesn't affect host application's implementations, I'm uncertain of the impact.

The risk is that it makes callers of `response.parsed_body` more flexible than the authors originally intended. 

While that's true, tests that are _currently_ passing would continue to pass, since String and Symbol keys are interchangeable. 

The trouble is that subsequent changes to those suites that used symbols would pass when the did not before. 

Some alternatives to this change:

* Wrap the `Hash` or `Array` constructed by `JSON.parse` in a new Class that **only** uses an indifferent access `Hash` when calling `#deconstruct_keys` to limit the change to **only** affect pattern matching
* Add a new [framework default](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#changing-behavior) for 7.1 that uses `object_class: ActiveSupport::HashWithIndifferentAccess`, then fall back to using `object_class: Hash`)
* revert any change to the implementation, and encourage callers to chain `.with_indifferent_access` or `.deep_symbolize_keys` when using Ruby's pattern matching